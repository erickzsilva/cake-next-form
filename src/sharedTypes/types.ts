export type FormInputs = {
  firstname: string
  lastName?: string
  deliveryDate?: Date
  deliveryTime?: Date
  phone: number
  email: string
  address1: string
  address2: string
  city: string
  region: string
  zipCode: string
  country: string
  redFruitCake: string
  strawberrySpongeCake: string
  hazelnutPie: string
  mmCake: string
}

export type CakeResponse = {
  title?: string
  description?: string
  textChooseYourCake: string
  subTitle: string
  cakes: Cakes[]
  country: string[]
}

export type Cakes = {
  value: string
  name: CakesFlavors
  alt: string
  srcImage: string
}

export interface NewCakes extends Cakes {
  paramValue: boolean
}

export type CakesFlavors =
  | 'redFruitCake'
  | 'strawberrySpongeCake'
  | 'hazelnutPie'
  | 'mmCake'