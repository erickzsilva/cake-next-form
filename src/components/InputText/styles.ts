import styled from 'styled-components'

export const Wrapper = styled.div`
  display: grid;
  position: relative;
`
export const Error = styled.span`
  color: red;
  font-size: 1.5rem;
`
export const Holder = styled.div`
  display: flex;
`
export const LabelField = styled.label`
  font-size: 11px;
  font-weight: 600;
  margin: 0 5px 4px 0;

  &.-error {
    color: #ff0101;
  }

  @media screen and (max-width: 768px) {
    &.-mr {
      margin: 0 5px 20px 0;
    }
  }
`
export const Input = styled.input`
  appearance: none;
  border: 1px solid #eba70e;
  border-radius: 3px;
  box-shadow: none;
  height: 35px;
  padding-left: 11px;
  -webkit-appearance: none;

  &.-util {
    color: #24457b;
    font-size: 11px;
    padding-left: 10px;
  }

  &::placeholder {
    color: #24457b;
    font-size: 11px;
  }
`
export const Info = styled.div`
  color: #bdbcc0;
  position: absolute;

  @media screen and (min-width: 768px) {
    bottom: -16px;
  }

  @media screen and (max-width: 768px) {
    top: 16px;
  }
`
