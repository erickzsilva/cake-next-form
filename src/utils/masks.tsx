type MaskType = 'zipCode' | 'onlyLetters' | 'onlyNumbers' | 'phone'

export const sendToFormatter = (value: string, maskType: MaskType) => {
  return objMasks[maskType](value)
}

const maskZipCode = (value: string) => {
  return value.replace(/\D/g, '').replace(/^(\d{5})(\d{3})+?$/, '$1-$2')
}

const maskOnlyLetters = (value: string) => {
  return value.replace(/[0-9!@#Â¨$%^&*)(+=._-]+/g, '')
}

const maskOnlyNumbers = (value: string) => {
  return value.replace(/\D/g, '')
}

const maskPhone = (value: string) => {
  return value
    .replace(/\D/g, '')
    .replace(/(\d{2})(\d)/, '($1) $2')
    .replace(/(\d{4})(\d)/, '$1-$2')
    .replace(/(\d{4})-(\d)(\d{4})/, '$1$2-$3')
    .replace(/(-\d{4})\d+?$/, '$1')
}

type ObjMasksProps = {
  zipCode: (value: string) => string
  onlyLetters: (value: string) => string
  onlyNumbers: (value: string) => string
  phone: (value: string) => string
}

const objMasks: ObjMasksProps = {
  zipCode: maskZipCode,
  onlyLetters: maskOnlyLetters,
  onlyNumbers: maskOnlyNumbers,
  phone: maskPhone
}
