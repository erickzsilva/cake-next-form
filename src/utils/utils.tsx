import { AppRouterInstance } from 'next/dist/shared/lib/app-router-context.shared-runtime'
import { ReadonlyURLSearchParams } from 'next/navigation'

import { CakesFlavors } from '../sharedTypes/types'

export const emailValidation: RegExp =
  /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/

export const testString = (value: string | null): boolean => {
  const result = value ? /true/.test(value) : false
  return result
}

export const cakesFlavors: CakesFlavors[] = [
  'redFruitCake',
  'strawberrySpongeCake',
  'hazelnutPie',
  'mmCake'
]

type MountParams = {
  name: string
  paramValue: boolean
  allParams: string
  router: AppRouterInstance
}

export const mountParams = ({
  name,
  paramValue,
  allParams,
  router
}: MountParams) => {
  cakesFlavors.forEach(() => {
    if (allParams) {
      if (allParams.includes(name)) {
        if (allParams.includes(`${name}=${paramValue}`)) {
          const newAllParams = allParams.replace(
            `${name}=${paramValue}`,
            `${name}=${!paramValue}`
          )
          router.replace(`?${newAllParams}`, {
            scroll: false
          })
        }
      } else {
        const newAllParams = allParams.concat(`&${name}=${!paramValue}`)
        router.replace(`?${newAllParams}`, {
          scroll: false
        })
      }
    } else {
      router.replace(`?${name}=${!paramValue}`, {
        scroll: false
      })
    }
  })
}

export const seachParams = (
  value: string,
  searchParams: ReadonlyURLSearchParams
): boolean => {
  const valParam = searchParams.get(value)
  return testString(valParam)
}

export const tomorrowDate = () => {
  const today = new Date()
  const tomorrow = new Date()

  tomorrow.setDate(today.getDate() + 1)
  tomorrow.setHours(0, 0, 0, 0)

  return tomorrow
    .toLocaleString('pt', {
      month: '2-digit',
      day: '2-digit',
      year: 'numeric'
    })
    .split('/')
    .reverse()
    .join('-')
}
