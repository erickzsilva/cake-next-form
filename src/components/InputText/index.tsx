import { HTMLInputTypeAttribute } from 'react'
import { FieldError, UseFormRegisterReturn } from 'react-hook-form'

import * as S from './styles'

export type InputProps = {
  autoComplete?: string
  label?: string
  maxLength?: number
  minDate?: string
  placeholder?: string
  type: HTMLInputTypeAttribute
  required?: FieldError
  register: UseFormRegisterReturn
}

const InputText = ({
  autoComplete,
  label,
  maxLength,
  minDate,
  placeholder,
  type,
  required,
  register
}: InputProps) => (
  <S.Wrapper>
    <S.Holder>
      <S.LabelField className={type === 'email' ? '-mr' : ''}>
        {label && <label htmlFor={register.name}>{label}</label>}
      </S.LabelField>
      <S.LabelField className="-error">
        {required && <S.Error>*</S.Error>}
      </S.LabelField>
    </S.Holder>
    <S.Input
      id={register.name}
      autoComplete={autoComplete}
      type={type}
      min={minDate}
      placeholder={placeholder}
      maxLength={maxLength}
      className={type === 'date' || type === 'time' ? '-util' : ''}
      {...register}
      data-testid="input"
    />

    {type === 'email' ? <S.Info>To receive the complete receipt</S.Info> : ''}
  </S.Wrapper>
)

export default InputText
