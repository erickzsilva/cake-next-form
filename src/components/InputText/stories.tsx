import { Meta } from '@storybook/react'
import InputText, { InputProps } from '.'
import { useForm } from 'react-hook-form'
import { FormInputs } from '@/sharedTypes/types'
import { sendToFormatter } from '../../utils/masks'
import { emailValidation, tomorrowDate } from '../../utils/utils'

export default {
  title: 'InputText',
  component: InputText
} as Meta<InputProps>

export const Default = () => {
  const {
    register,
    formState: { errors }
  } = useForm<FormInputs>()

  return (
    <>
      <div>
        <InputText
          placeholder="First"
          register={register('firstname', {
            required: true,
            maxLength: 20,
            pattern: /^[A-Za-z]+$/i,
            onChange(event) {
              const { value } = event.target
              event.target.value = sendToFormatter(value, 'onlyLetters')
            }
          })}
          label="Name"
          type="text"
          maxLength={20}
          autoComplete={'given-name'}
          required={errors.firstname}
        />
      </div>

      <div>
        <InputText
          placeholder="Last"
          register={register('lastName', {
            maxLength: 20,
            onChange(event) {
              const { value } = event.target
              event.target.value = sendToFormatter(value, 'onlyLetters')
            }
          })}
          type="text"
          maxLength={20}
          autoComplete={'family-name'}
        />
      </div>

      <div>
        <InputText
          label="Delivery date"
          minDate={tomorrowDate()}
          register={register('deliveryDate')}
          type="date"
        />
      </div>

      <div>
        <InputText
          label="Preferred delivery time"
          register={register('deliveryTime')}
          type="time"
        />
      </div>

      <div>
        <InputText
          placeholder="### ### ####"
          register={register('phone', {
            required: true,
            onChange(event) {
              const { value } = event.target
              event.target.value = sendToFormatter(value, 'phone')
            }
          })}
          label="Phone"
          type="tel"
          required={errors.phone}
          autoComplete="tel-local"
        />
      </div>

      <div>
        <InputText
          placeholder=""
          register={register('email', {
            required: true,
            pattern: emailValidation
          })}
          label="Email"
          type="text"
          required={errors.email}
          autoComplete="email"
        />
        <p>To receive the complete receipt</p>
      </div>

      <div>
        <InputText
          placeholder="Street Address"
          register={register('address1')}
          maxLength={80}
          autoComplete="address-line1"
          type="text"
        />
      </div>

      <div>
        <InputText
          placeholder="Street Address Line 2"
          register={register('address2')}
          maxLength={80}
          autoComplete="address-line2"
          type="text"
        />
      </div>
      <div>
        <InputText
          placeholder="City"
          register={register('city')}
          maxLength={20}
          autoComplete="address-level1"
          type="text"
        />
      </div>
      <div>
        <InputText
          placeholder="Region"
          register={register('region')}
          maxLength={20}
          autoComplete="address-level2"
          type="text"
        />
      </div>
      <div>
        <InputText
          placeholder="Postal / Zip Code"
          register={register('zipCode', {
            onChange(event) {
              const { value } = event.target
              event.target.value = sendToFormatter(value, 'zipCode')
            }
          })}
          type="tel"
          autoComplete="postal-code"
        />
      </div>
      <div></div>
      <div></div>
    </>
  )
}
