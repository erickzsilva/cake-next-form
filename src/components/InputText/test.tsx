import { render, screen } from '@testing-library/react'
import { renderHook } from '@testing-library/react'

import InputText from '.'
import { SubmitHandler, useForm } from 'react-hook-form'
import { FormInputs } from '../../sharedTypes/types'
import { sendToFormatter } from '../../utils/masks'

describe('<InputText />', () => {
  it('should render the an input text', () => {
    const { result } = renderHook(() => useForm<FormInputs>())
    const onSubmit: SubmitHandler<FormInputs> = () => ''

    const { container } = render(
      <form onSubmit={result.current.handleSubmit(onSubmit)}>
        <InputText
          placeholder="First"
          register={result.current.register('firstname', {
            required: true,
            maxLength: 20,
            pattern: /^[A-Za-z]+$/i,
            onChange(event) {
              const { value } = event.target
              event.target.value = sendToFormatter(value, 'onlyLetters')
            }
          })}
          label="Name"
          type="text"
          maxLength={20}
          autoComplete={'given-name'}
          required={result.current.formState.errors.firstname}
        />
        <input value="Order" type="submit" />
      </form>,
      {}
    )

    expect(screen.getByTestId('input')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
})
