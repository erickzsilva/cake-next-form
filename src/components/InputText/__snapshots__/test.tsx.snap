// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputText /> should render the an input text 1`] = `
.c0 {
  display: grid;
  position: relative;
}

.c1 {
  display: flex;
}

.c2 {
  font-size: 11px;
  font-weight: 600;
  margin: 0 5px 4px 0;
}

.c2.-error {
  color: #ff0101;
}

.c3 {
  appearance: none;
  border: 1px solid #eba70e;
  border-radius: 3px;
  box-shadow: none;
  height: 35px;
  padding-left: 11px;
  -webkit-appearance: none;
}

.c3.-util {
  color: #24457b;
  font-size: 11px;
  padding-left: 10px;
}

.c3::placeholder {
  color: #24457b;
  font-size: 11px;
}

@media screen and (max-width: 768px) {
  .c2.-mr {
    margin: 0 5px 20px 0;
  }
}

<form>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        class="c2"
      >
        <label
          for="firstname"
        >
          Name
        </label>
      </label>
      <label
        class="c2 -error"
      />
    </div>
    <input
      autocomplete="given-name"
      class="c3"
      data-testid="input"
      id="firstname"
      maxlength="20"
      name="firstname"
      placeholder="First"
      type="text"
    />
  </div>
  <input
    type="submit"
    value="Order"
  />
</form>
`;
